### PR TITLE
Tag filter CLI

### PR DESCRIPTION
**Pull request by sambooo**
Currently working towards implementing this command:

```
$ fluxctl policy --service=foo \
    [--(de)automate] \
    [--(un)lock] \
    [--container=bar --tag=master-*]
```

Progress:

- [x] `--(de)automate`
- [x] `--(un)lock`
- [x] `--container --tag` (with arbitrarily many pairs of these)
- [x] commit messages must make sense
- [x] history

### CODE REVIEW

**Comment by sambooo** [comment]
Does it make sense to instead make subcommands under policy? ie:

```
$ fluxctl policy automate --service=foo
$ fluxctl policy filter --service=foo --container=bar --tag=baz
$ fluxctl policy lock --service=foo
```

**Comment by squaremo** [comment]
> Does it make sense to instead make subcommands under policy

Not if you want to do more than one at a time, e.g., automate with a particular tag. You'd have to have a (more of) a model in your head about how to get where you want to go -- e.g., don't automate before supplying the filter.

**Comment by sambooo** [comment]
Good point. A user that knows they want to both automate and specify a tag shouldn't need to figure out which order to do those actions in.

**Comment by squaremo** [comment]
> --container --tag

Will these come in pairs? E.g., if you want to set the filter on more than one container at a time.

If you're automating a service that hasn't been automated before, you might want to provide filters for each container, so that you don't accidentally let your sidecar image roam free.

**Comment by sambooo** [comment]
Good point about sidecar! Without allowing multiple pairs to be specified we'd end up with pretty much the same issue as above.

I shall look into how to make this work

File: cmd/fluxctl/policy.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	if opts.deautomate {
+		remove = remove.Add(policy.Automated)
+	}
+	if opts.unlock {
+		remove = remove.Add(policy.Locked)
+	}
+
+	for i, container := range opts.containers {
+		tag := opts.tags[i]
+		if tag != "" && tag != "*" {
```
Comment (philwinder) [inline cmd/fluxctl/policy.go]: Is `--tag=""` some special case? I don't think that would be allowed by the cli argument parser.

File: cmd/fluxctl/policy.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			add = add.Set(policy.TagPrefix(container), "glob:"+tag)
+		} else if tag == "*" {
+			remove = remove.Add(policy.TagPrefix(container))
+		}
+	}
+
+	update := policy.Update{}
+	if len(add) > 0 {
+		update.Add = add
```
Comment (philwinder) [inline cmd/fluxctl/policy.go]: `update` will have an empty `Add` and `Remove`. So there's no need for the check. It's safe to set `update.Add` to another empty add (same for remove).

File: daemon/daemon.go [hunk 3]
```diff
@@ -455,20 +455,24 @@ func policyCommitMessage(us policy.Updates, cause update.Cause) string {
 func policyEventTypes(u policy.Update) []string {
 	types := map[string]struct{}{}
```
Comment (philwinder) [inline daemon/daemon.go]: I don't see the point of using an empty struct here. Why not just an array?

Just looked at the rest of the code. It eventually returns an array anyway, so just swap this out to be an array.

Comment (sambooo) [inline daemon/daemon.go]: This is a common Go pattern used to simulate a Set. Without this we'd end up with duplicates.

**Comment by sambooo** [comment]
@philwinder Thanks for the review!


The globbing just uses the go-glob library which allows arbitrarily many asterisks in its patterns.

Do you think that we need tests to check that the library we use does what it should? I'm not sure whether that's necessary but if you think so then I'll write some.

**Comment by sambooo** [comment]
After some discussion with Michael, the command is being changed to this:

```
$ fluxctl policy --service=foo \
    [--(de)automate] \
    [--(un)lock] \
    [--tag-all='*'] `
    [--tag='container=pattern']+
```

Where the `--tag` flags are applied after `--tag-all`, so that a default can be specified and overridden in a single command.

File: cmd/fluxctl/policy.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		remove = remove.Add(policy.Automated)
+	}
+	if opts.unlock {
+		remove = remove.Add(policy.Locked)
+	}
+
+	for _, tagPair := range opts.tags {
+		parts := strings.Split(tagPair, "=")
+		if len(parts) != 2 {
+			return policy.Update{}, fmt.Errorf("invalid container/tag pair: %q", tagPair)
```
Comment (squaremo) [inline cmd/fluxctl/policy.go]: Perhaps spell out what the expected format is, here.

File: history/event.go [hunk 5]
```diff
@@ -23,6 +23,7 @@ const (
 	EventDeautomate  = "deautomate"
 	EventLock        = "lock"
 	EventUnlock      = "unlock"
+	EventUpdateTag   = "update_tag"
```
Comment (squaremo) [inline history/event.go]: I foresee there being more kinds of policy, and we could get bogged down in individual events for each. How about we have a PolicyChange event which subsumes the others, and has the previous and new policies. We can attach labels, if necessary, to make it easier to tell whether it represents a transition e.g., locked, automated.

Comment (squaremo) [inline history/event.go]: (Yes we'll have to make compensating changes in the UI)

**Pull request merged**